src/app.js

```diff
@@ -221,30 +216,128 @@ export default function (app) {
 
   function patch(parent, element, oldNode, node) {
     if (oldNode == null) {
-      element = parent.appendChild(createElementFrom(node))
+      element = parent.insertBefore(createElementFrom(node), element)
 
-    } else if (node == null) {
-      batch.push(function () {
-        if (oldNode.data && oldNode.data.onRemove) {
-          oldNode.data.onRemove(element)
-        }
-        parent.removeChild(element)
-      })
     } else if (node.tag && node.tag === oldNode.tag) {
-      updateElementData(element, node.data, oldNode.data)
+      updateElementData(element, oldNode.data, node.data)
 
       var len = node.children.length
       var oldLen = oldNode.children.length
+      var reusableChildren = {}
+      var oldElements = []
+      var newKeys = {}
+
+
+      for (var i = 0; i < oldLen; i++) {
+        var oldElement = element.childNodes[i]
+        oldElements[i] = oldElement
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
+        var oldChild = oldNode.children[i]
+        var oldKey = getKeyFrom(oldChild)
+
+        if (null != oldKey) {
+          reusableChildren[oldKey] = [oldElement, oldChild]
+        }
       }
 
+      var i = 0
```

Comment:
Weird? You haven't seen weird. I just finished banging my head against a performance regression caused by Google Closure Compiler inclining a function into a parent function because it became smaller - "this will be faster", i thought.

It did this to reduce size, okay. But when it came to benchmarks, the Chrome's JIT wasn't fastpathing the parent now because the entire parent became larger as a result of the inlining. I had to hack in a thing to prevent it from inlining it to get the perf back (at the expense of 20 bytes.)

[1] https://github.com/leeoniya/domvm/commit/e3a02fc9e301e0c038f8ac4505a93d5846a8fb64